include/seqan3/io/stream/iterator.hpp

```diff
@@ -166,6 +172,204 @@ class fast_istreambuf_iterator
+    fast_ostreambuf_iterator & operator=(char_t const c)
+    {
+        assert(stream_buf != nullptr);
+        if (stream_buf->pptr() == stream_buf->epptr())
+        {
+            if (stream_buf->sputc(c) == traits_t::eof()) // overflow() [virtual], then write character
+                throw std::ios_base::failure{"Cannot write to output stream (reached traits::eof() condition)."};
+        }
+        else
+        {
+            *stream_buf->pptr() = c;
+            stream_buf->pbump(1); // advance pptr() in put area without any checks
+        }
+        return *this;
+    }
+
+    //!\brief Returns `true if this iterator has encountered the end-of-file condition on output, `false` otherwise.
+    bool failed() const noexcept
+    {
+        return stream_buf->overflow() == traits_t::eof();
+    }
+
+    /*!\brief Writes a range to the associated output.
+     * \tparam range_type The type of range to write; Must model std::ranges::forward_range.
+     * \param[in] rng The range to write.
+     * \returns If `range_type` models `std::ranges::forwarding_range` returns an iterator pointing to end of the range
+     *          (rng) else returns `void`.
+     *
+     * This function avoids the buffer-at-end check by writing the range in chunks, where a chunks has the size of
+     * the remaining space in the put area of the buffer.
+     * If the range type models `std::ranges::sized_range` the chunks are written using `std::ranges::copy_n`, which
+     * may use memcpy if applicable. Otherwise, a simple for loop iterates over the chunk.
+     *
+     * \attention You can only use the return value (end iterator) if your range type models
+     *            `std::ranges::forwarding_range`.
+     *
+     * Example:
+     *
+     * \include test/snippet/io/detail/iterator_write_range.cpp
+     */
+    template <std::ranges::forward_range range_type>
+    //!\cond
+        requires forwarding_range<range_type>
+    //!\endcond
+    auto write_range(range_type && rng)
+    {
+        using sen_t = std::ranges::sentinel_t<range_type>;
+        using it_t = std::ranges::iterator_t<range_type>;
+
+        it_t it = std::ranges::begin(rng);
+        sen_t end = std::ranges::end(rng);
+
+        while (it != end)
+        {
+            size_t const buffer_space = stream_buf->epptr() - stream_buf->pptr();
+
+            if constexpr (std::ranges::sized_range<range_type>)
+            {
+                size_t const characters_to_write = std::min<size_t>(std::ranges::distance(it, end), buffer_space);
+                auto copy_res = std::ranges::copy_n(it, characters_to_write, stream_buf->pptr());
+                it = copy_res.in;
+                stream_buf->pbump(characters_to_write);
+            }
+            else
+            {
+                for (size_t i = 0; it != end && i < buffer_space; ++it, ++i)
+                {
+                    *stream_buf->pptr() = *it;
+                    stream_buf->pbump(1);
+                }
```

Comment:
```cpp
                size_t i = 0;
                for (; it != end && i < buffer_space; ++it, ++i)
                    *stream_buf->pptr() = *it;

                stream_buf->pbump(i);
```

Only one virtual function call?

Reply:
pbump is not a virtual function call!

but it might be faster nonetheless